### PR TITLE
Adding an e2e test on Windows GMSA support

### PR DESCRIPTION
**Pull request by wk8**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This patch is adding an e2e on the Windows GMSA support added in
https://github.com/kubernetes/kubernetes/pull/73726 and
https://github.com/kubernetes/kubernetes/pull/74737.

**Which issue(s) this PR fixes**:
@yujuhong requested an e2e test on this at https://github.com/kubernetes/kubernetes/pull/73726#pullrequestreview-206595783.

**Special notes for your reviewer**:
This e2e test has been verified to pass on a cluster with a Windows node with the
`WindowsGMSA=true` feature gate enabled (and the kubelet compiled from
https://github.com/kubernetes/kubernetes/pull/74737); however, I don't believe that
feature gate is enabled when running a "regular" build. Would it be possible to:
1) enable that feature gate when running Windows e2e tests in a build?
2) or else enable dynamic Kubelet config for these tests, so that each spec can
set the config it needs?

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by wk8** [comment]
@yujuhong @dchen1107 @michmike @PatrickLang : guidance appreciated as to how to make sure that e2e test can run successfully as part of builds (see the `Special notes for your reviewer` section in the  PR message). Thanks!

File: test/e2e/windows/density.go [hunk 1]
```diff
@@ -30,21 +30,14 @@ import (
...
 )
 
 var _ = SIGDescribe("[Feature:Windows] Density [Serial] [Slow]", func() {
-
 	f := framework.NewDefaultFramework("density-test-windows")
 
-	BeforeEach(func() {
-		// NOTE(vyta): these tests are Windows specific
-		framework.SkipUnlessNodeOSDistroIs("windows")
-	})
```
Comment (wk8) [inline test/e2e/windows/density.go]: Note to reviewers: this has been moved to https://github.com/kubernetes/kubernetes/pull/74738/files#diff-b0ddb1c276f901c22122f6283cc7348c

**Review by neolit123 — COMMENTED** [review]
/area test
/kind feature
/priority important-soon

**Comment by michmike** [comment]
@adelina-t  and @bclau  can you please read the comments above regarding how to light up this test in automation

**Comment by michmike** [comment]
@neolit123 can you also please review the test?

File: test/e2e/windows/gmsa.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// We need an image that packs the `nltest` command for this - and the `winPauseImageName` one doesn't.
+// I'm afraid this is the smallest one that does.
+const gMSATestImageName = "mcr.microsoft.com/powershell"
+
+var _ = SIGDescribe("[Feature:Windows] GMSA [Slow]", func() {
```
Comment (ddebroy) [inline test/e2e/windows/gmsa.go]: The Feature tag above, I think, should also refer to the GMSA feature tag.

File: test/e2e/windows/gmsa.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// We need an image that packs the `nltest` command for this - and the `winPauseImageName` one doesn't.
+// I'm afraid this is the smallest one that does.
+const gMSATestImageName = "mcr.microsoft.com/powershell"
```
Comment (neolit123) [inline test/e2e/windows/gmsa.go]: yeah too bad nltest is not part of the nanoserver images...
but also nl*.dll from system32 isn't either.

File: test/e2e/windows/gmsa.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+var _ = SIGDescribe("[Feature:Windows] GMSA [Slow]", func() {
+	f := framework.NewDefaultFramework("gmsa-test-windows")
+
+	Describe("kubelet GMSA support", func() {
+		Context("when creating a pod with correct GMSA credential specs", func() {
+			It("passes the credential specs down to the pod's containers", func() {
```
Comment (neolit123) [inline test/e2e/windows/gmsa.go]: non blocking nit: ideally `pod`-> `Pod` in all user facing strings.

**Review by neolit123 — COMMENTED** [review]
LGTM
as much as i understand about nltest.exe and the AD config that is tested here.

File: test/e2e/windows/gmsa.go (hunk 2 above)
Comment (wk8) [reply test/e2e/windows/gmsa.go]: 👍

File: test/e2e/windows/gmsa.go (hunk 4 above)
Comment (wk8) [reply test/e2e/windows/gmsa.go]: 👍

**Review by michmike — APPROVED** [review]

File: test/e2e/windows/framework.go [hunk 5]
```diff
@@ -16,8 +16,23 @@ limitations under the License.
...
-import "github.com/onsi/ginkgo"
+import (
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+)
+
+// it seems we can't use k8s.io/kubernetes/test/utils/image.GetPauseImageName() as it does not
+// contain a windows image?
+const winPauseImageName = "kubeletwin/pause:latest"
```
Comment (adelina-t) [inline test/e2e/windows/framework.go]: Not a fan of this being hardcoded here.

Also, the pause image here ( the one returned by GetPauseImageName() ): https://github.com/kubernetes/kubernetes/blob/b67126e350fdba6d89a0ee64310cac7db14e33d3/test/utils/image/manifest.go#L128 can be a windows image. 

By passing this env var: https://github.com/kubernetes/kubernetes/blob/b67126e350fdba6d89a0ee64310cac7db14e33d3/test/utils/image/manifest.go#L66 

with a file containing the registry names for windows images, the pause image should be fine.

In the upstream tests we use these registries: https://github.com/kubernetes-sigs/windows-testing/blob/master/images/image-repo-list

In fact, all the dockerfiles + documentation of the images we use in upstream e2e windows tests is here:

https://github.com/kubernetes-sigs/windows-testing

File: test/e2e/windows/gmsa.go (hunk 3 above)
Comment (adelina-t) [reply test/e2e/windows/gmsa.go]: Same comment as above. No need to hardcore. We should just use one of the images in "k8s.io/kubernetes/test/utils/image", they are all based on windowsservercore , thus will have powershell. If not, we can just define another image in "k8s.io/kubernetes/test/utils/image", either way I suggest we stick to having all image names and repos there.

File: test/e2e/windows/density.go [hunk 6]
```diff
@@ -30,21 +30,14 @@ import (
 	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/kubernetes/test/e2e/framework"
-	imageutils "k8s.io/kubernetes/test/utils/image"
```
Comment (adelina-t) [inline test/e2e/windows/density.go]: All image names should be imported from here for consistency

**Review by adelina-t — COMMENTED** [review]
Thank you for this PR :). Left some comments inside.

File: test/e2e/windows/gmsa.go (hunk 3 above)
Comment (wk8) [reply test/e2e/windows/gmsa.go]: Neat, hadn't seen the trick of overriding the repos. Thanks.

File: test/e2e/windows/density.go (hunk 6 above)
Comment (wk8) [reply test/e2e/windows/density.go]: 👍

File: test/e2e/windows/framework.go (hunk 5 above)
Comment (wk8) [reply test/e2e/windows/framework.go]: 👍

**Comment by wk8** [comment]
Thanks for the review @adelina-t  :) addressed your comments.

File: test/e2e/windows/README.md [hunk 7]
```diff
@@ -4,7 +4,7 @@
 
 ```bash
 KUBECONFIG=path/to/kubeconfig
-curl https://raw.githubusercontent.com/e2e-win/e2e-win-prow-deployment/master/repo-list -o repo_list
+curl https://raw.githubusercontent.com/e2e-win/e2e-win-prow-deployment/master/repo-list -o repo_list # or https://raw.githubusercontent.com/kubernetes-sigs/windows-testing/master/images/image-repo-list
```
Comment (bclau) [inline test/e2e/windows/README.md]: Considering that kubernetes-sigs/windows-testing is the official repo, that would make it's image-repo-list file official and up to date. I would just have that file mentioned, no need for both.

File: test/e2e/windows/gmsa.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+				container2Name := "container2"
+				container2Domain := "contoso.org"
+
+				containers := make([]corev1.Container, 2)
+				for i, name := range []string{container1Name, container2Name} {
+					containers[i] = corev1.Container{
+						Name:    name,
+						Image:   imageutils.GetPauseImageName(),
+						Command: []string{"powershell"},
+						Args:    []string{"while ($true) { Start-Sleep -Seconds 1 }"},
```
Comment (bclau) [inline test/e2e/windows/gmsa.go]: Well, the command and args are unnecessary. It is a pause image after all, it is supposed to do that anyways.

File: test/e2e/windows/README.md (hunk 7 above)
Comment (wk8) [reply test/e2e/windows/README.md]: 👍

File: test/e2e/windows/gmsa.go (hunk 8 above)
Comment (wk8) [reply test/e2e/windows/gmsa.go]: Good catch, thanks. Changed.

Sorry I forgot changing this as part of 9eac1bebd024ccf510ab126c08d7dfacd1e651d0 (was using image `mcr.microsoft.com/powershell` before).

File: test/e2e/windows/README.md [hunk 9]
```diff
@@ -4,7 +4,7 @@
 
 ```bash
 KUBECONFIG=path/to/kubeconfig
-curl https://raw.githubusercontent.com/e2e-win/e2e-win-prow-deployment/master/repo-list -o repo_list
+curl https://raw.githubusercontent.com/kubernetes-sigs/windows-testing/master/images/image-repo-list -o repo_list
```
Comment (ddebroy) [inline test/e2e/windows/README.md]: Can this be moved to a separate commit from the main GMSA test commit?

File: test/e2e/windows/gmsa.go [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+					if !strings.Contains(output, expectedSubstr) {
+						framework.Failf("Expected %q to contain %q", output, expectedSubstr)
+					}
+				}
+
+				// If this was an e2e_node test, we could also check that the registry keys used to pass down the cred specs to Docker
+				// have been properly cleaned up - but as of right now, e2e_node tests don't support Windows. We should migrate this
+				// test to an e2e_node test when they start supporting Windows.
+
+				// FIXME: remove this before merging - this is here to check if this test is actually run
```
Comment (ddebroy) [inline test/e2e/windows/gmsa.go]: This needs to be removed/fixed right?

**Review by ddebroy — COMMENTED** [review]
Looks good overall. Please address the FIX comment, squash the commits and separate out the README fix to a separate commit.

File: test/e2e/windows/gmsa.go (hunk 10 above)
Comment (wk8) [reply test/e2e/windows/gmsa.go]: Yes - but before I remove it I'd like to see it failing where it's supposed to be running, to make sure this test actually runs somewhere... And I haven't seen that yet?

File: test/e2e/windows/README.md (hunk 9 above)
Comment (wk8) [reply test/e2e/windows/README.md]: 👍

File: test/e2e/windows/gmsa.go (hunk 10 above)
Comment (ddebroy) [reply test/e2e/windows/gmsa.go]: Discussed with @wk8 offline that these tests won't run as PROW or get picked up by the other testgrid suites right now.

**Review by michmike — APPROVED** [review]

File: test/e2e/windows/gmsa.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (yujuhong) [inline test/e2e/windows/gmsa.go]: s/2018/2019

File: test/e2e/windows/gmsa.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+					Spec: corev1.PodSpec{
+						Containers: containers,
+					},
+				}
+
+				ginkgo.By("creating a pod with correct GMSA annotations")
+				f.PodClient().Create(pod)
+
+				ginkgo.By("waiting for the pod and its containers to be running")
+				gomega.Eventually(func() bool {
```
Comment (yujuhong) [inline test/e2e/windows/gmsa.go]: You can just use `framework.WaitForPodReady()`
There are other utilities functions  like that. Is there any reason to roll your own?

Comment (wk8) [reply test/e2e/windows/gmsa.go]: No reason other than I didn't spot these functions, apologies. @adelina-t kindly fixed that in https://github.com/kubernetes/kubernetes/pull/75116
